REFERENCE.md

```diff
@@ -0,0 +1,1404 @@
+`chain` or `jump` parameters, the firewall resource will autorequire
+those firewallchain resources.
+
+If Puppet is managing the iptables, iptables-persistent, or iptables-services packages,
+and the provider is iptables or ip6tables, the firewall resource will
+autorequire those packages to ensure that any required binaries are
+installed.
+
+#### Properties
+
+The following properties are available in the `firewall` type.
+
+##### `ensure`
+
+Valid values: present, absent
+
+Manage the state of this rule. The default action is *present*.
+
+Default value: present
+
+##### `action`
+
+Valid values: accept, reject, drop
+
+This is the action to perform on a match. Can be one of:
+
+* accept - the packet is accepted
+* reject - the packet is rejected with a suitable ICMP response
+* drop - the packet is dropped
+
+If you specify no value it will simply match the rule but perform no
+action unless you provide a provider specific parameter (such as *jump*).
+
+##### `source`
+
+The source address. For example:
+
+    source => '192.168.2.0/24'
+
+You can also negate a mask by putting ! in front. For example:
+
+    source => '! 192.168.2.0/24'
+
+The source can also be an IPv6 address if your provider supports it.
+
+##### `src_range`
+
+The source IP range. For example:
+
+    src_range => '192.168.1.1-192.168.1.10'
+
+The source IP range must be in 'IP1-IP2' format.
+
+##### `destination`
+
+The destination address to match. For example:
+
+    destination => '192.168.1.0/24'
+
+You can also negate a mask by putting ! in front. For example:
+
+    destination  => '! 192.168.2.0/24'
+
+The destination can also be an IPv6 address if your provider supports it.
+
+##### `dst_range`
+
+The destination IP range. For example:
+
+    dst_range => '192.168.1.1-192.168.1.10'
+
+The destination IP range must be in 'IP1-IP2' format.
+
+##### `sport`
+
+The source port to match for this filter (if the protocol supports
+ports). Will accept a single element or an array.
+
+For some firewall providers you can pass a range of ports in the format:
+
+    <start_number>-<ending_number>
+
+For example:
+
+    1-1024
+
+This would cover ports 1 to 1024.
+
+##### `dport`
+
+The destination port to match for this filter (if the protocol supports
+ports). Will accept a single element or an array.
+
+For some firewall providers you can pass a range of ports in the format:
+
+    <start_number>-<ending_number>
+
+For example:
+
+    1-1024
+
+This would cover ports 1 to 1024.
+
+##### `port`
+
+DEPRECATED
+
+The destination or source port to match for this filter (if the protocol
+supports ports). Will accept a single element or an array.
+
+For some firewall providers you can pass a range of ports in the format:
+
+    <start_number>-<ending_number>
+
+For example:
+
+    1-1024
+
+This would cover ports 1 to 1024.
+
+##### `dst_type`
+
+Valid values: [:UNSPEC, :UNICAST, :LOCAL, :BROADCAST, :ANYCAST, :MULTICAST,
```

Comment:
Formatting broken here.

Reply:
This is taken from code, not documentation. Not something I can format.
```
newproperty(:proto) do
    desc <<-PUPPETCODE
      The specific protocol to match for this rule.
    PUPPETCODE

    newvalues(*[:ip, :tcp, :udp, :icmp, :"ipv6-icmp", :esp, :ah, :vrrp, :igmp, :ipencap, :ipv4, :ipv6, :ospf, :gre, :cbt, :sctp, :pim, :all].map { |proto|
      [proto, "! #{proto}".to_sym]
    }.flatten)
    defaultto 'tcp'
  end
```